registry/registry.go

```diff
@@ -103,13 +104,37 @@ func (r *RouteRegistry) Register(uri route.Uri, endpoint *route.Endpoint) {
 }
 
 func (r *RouteRegistry) register(uri route.Uri, endpoint *route.Endpoint) route.PoolPutResult {
-	r.Lock()
-	defer r.Unlock()
+	r.RLock()
+	defer r.RUnlock()
 
 	t := time.Now()
-
 	routekey := uri.RouteKey()
+	pool := r.byURI.Find(routekey)
+
+	if pool == nil {
+		// release read lock, insertRouteKey() will acquire a write lock.
+		r.RUnlock()
+		pool = r.insertRouteKey(routekey, uri)
+		r.RLock()
+	}
+
+	if endpoint.StaleThreshold > r.dropletStaleThreshold || endpoint.StaleThreshold == 0 {
+		endpoint.StaleThreshold = r.dropletStaleThreshold
+	}
+
+	endpointAdded := pool.Put(endpoint)
+
+	r.SetTimeOfLastUpdate(t)
+
+	return endpointAdded
+}
+
+// insertRouteKey acquires a write lock, inserts the route key into the registry and releases the write lock.
+func (r *RouteRegistry) insertRouteKey(routekey route.Uri, uri route.Uri) *route.EndpointPool {
+	r.Lock()
+	defer r.Unlock()
 
+	// double check that the route key is still not found, now with the write lock.
```

Comment:
the function call to `r.byURI.Find(routekey)` is pretty expensive as it traverses through the Trie structure sequentially. If possible I'd like to avoid doing that twice. In what scenario could there have been a change to the Trie between here and L112? While you're doing the first lookup you already have a read-lock obtained, so no one else could have acquired a write-lock and have inserted an element at the `routekey` location that would conflict? so why do the lookup twice?

Reply:
It's an instance of [Double Checked Locking](https://en.wikipedia.org/wiki/Double-checked_locking).

A process or thread can be interrupted/preempted at pretty much any time, so there is a slim chance that we look up reading, release the lock, and before _we_ acquire the write lock some other thread may have acquired it and changed the value. And _then_ we get the write lock and work on outdated assumptions.

Without this double check, there is a low-probability race condition that this double check avoids entirely.

The lookup of `r.byURI.Find()` is reasonably trivial. The route URI is split by `/` and those segments are iterated over.
This is actually also improved further by #379, by eliminating the alloc caused by the `parts()` function.

Reply:
> eliminating the alloc caused by the parts() function

I was worried about that exact part 😁
Thanks for pointing to the other fix in that area.